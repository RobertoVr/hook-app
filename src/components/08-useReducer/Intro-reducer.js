
const initilState = [{
    id: 1,
    todo: 'Comer pan',
    done: false
}];

export const ToDoReducer = (state = initilState, action) => {
    if (action?.type === 'agregar') {
        return [...state, action.payload];
    }
    return state;
}

const toDo = ToDoReducer();

const newToDo = {
    id: 2,
    todo: 'Comprar leche',
    done: false
}

const action = {
    type: 'agregar',
    payload: newToDo
}

const res = ToDoReducer(toDo, action);


console.log(res);