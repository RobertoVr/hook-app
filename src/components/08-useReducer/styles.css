body {
  padding: 50px;
}

p {
  user-select: none;
  cursor: pointer;
  margin: 0px;
}

.list-group-item {
  align-items: center;
  justify-content: space-between;
  display: flex;
}

.complete {
  text-decoration: line-through;
}
