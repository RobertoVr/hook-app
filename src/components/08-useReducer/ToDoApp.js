import React, { useEffect, useReducer } from 'react'
import { todoReducer } from './todoReducer';
import { TodoList } from './todoList/TodoList';
import { TodoAdd } from './form/TodoAdd';

import './styles.css'

const init = () => {
    return JSON.parse(localStorage.getItem('todos')) || [];
    // return [{
    //     id: new Date().getTime(),
    //     desc: 'Aprender React',
    //     done: false,
    // }]
}

export const ToDoApp = () => {

    const [todos, dispatch] = useReducer(todoReducer, [], init);


    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(todos));
    }, [todos]);

    const handleDelete = (todoId) => {
        const action = {
            type: 'delete',
            payload: todoId
        }
        dispatch(action);
    }

    const handleToggle = (todoId) => {
        dispatch({
            type: 'toggle',
            payload: todoId
        });
    }

    const handleAddToDo = (newTodo) => {
        dispatch({
            type: 'add',
            payload: newTodo
        });
    }

    return (
        <div>
            <h1>To Do App ({todos.length})</h1>
            <div className="row">
                <div className="col-7">
                    {/* TodoList, todos prop, handleDelete, handleToggle */}
                    <TodoList
                        todos={todos}
                        handleDelete={handleDelete}
                        handleToggle={handleToggle}
                    />
                </div>
                <div className="col-5">
                    <TodoAdd
                        handleAddToDo={handleAddToDo}
                    />
                </div>
            </div>
        </div>
    )
}
