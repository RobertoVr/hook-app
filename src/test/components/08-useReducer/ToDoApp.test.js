import React from 'react';
import '@testing-library/jest-dom';
import { mount, shallow } from 'enzyme';
import { act } from '@testing-library/react';
import { ToDoApp } from '../../../components/08-useReducer/ToDoApp';
import { demoTodos } from '../../fixtures/demoTodos';

describe('Pruebas en <ToDoApp></ToDoApp>', () => {

    const wrapper = shallow(<ToDoApp />);
    Storage.prototype.setItem = jest.fn(() => { });

    test('Debe de mostrar el componente', () => {
        expect(wrapper).toMatchSnapshot();
    });
    // test('Debe de agregar un TODO', () => {
    //     const wrapper = mount(<ToDoApp />);
    //     act(() => {
    //         wrapper.find('TodoAdd').prop('handleAddToDo')(demoTodos[0]);
    //         wrapper.find('TodoAdd').prop('handleAddToDo')(demoTodos[1]);
    //     });
    //     expect(wrapper.find('h1').text().trim()).toBe('To Do App (2)');
    //     expect(localStorage.setItem).toHaveBeenCalledTimes(2);
    // });
    test('debe eliminar un todo', () => {
        wrapper.find('TodoAdd').prop('handleAddToDo')(demoTodos[0]);
        expect(wrapper.find('h1').text().trim()).toBe('To Do App (1)');
        wrapper.find('TodoList').prop('handleDelete')(demoTodos[0].id);
        expect(wrapper.find('h1').text().trim()).toBe('To Do App (0)');
    })
    
});
