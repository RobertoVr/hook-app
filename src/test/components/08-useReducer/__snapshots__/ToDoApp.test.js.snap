// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <ToDoApp></ToDoApp> Debe de mostrar el componente 1`] = `
<div>
  <h1>
    To Do App (
    0
    )
  </h1>
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <TodoList
        handleDelete={[Function]}
        handleToggle={[Function]}
        todos={Array []}
      />
    </div>
    <div
      className="col-5"
    >
      <TodoAdd
        handleAddToDo={[Function]}
      />
    </div>
  </div>
</div>
`;
